### ISSUE TITLE
Relax attribute lookup in warning about old-styled middleware

### ISSUE THREAD

**Issue by asvetlov** [aio-libs/aiohttp#2340]
For class-based middleware like
```
class Middleware:
    async def __call__(self, app, handler):
          ...
```
there is no `__name__` attr.
Moreover I could imaging other situations where `__name__` doesn't exists.
It breaks web server starting.

Let's change `middleware.__name__` to `repr(middleware)`.

Bugfix should be done in `2.3` branch.

**Comment by samuelcolvin** [comment]
do you want me to do this?

**Comment by asvetlov** [comment]
Could you do it asap? Otherwise I'll make a fix myself.
Because I'm blocked on switching to new aiohttp release.
Another finding: even if I decorated the method the check doesn't work:
```
class Middleware:
    @web.middleware
    async def __call__(self, app, handler):
          ...
```
`_prepare_middlewares()` should look on `__func__` in case of methoddescriptor.

**Comment by samuelcolvin** [comment]
Sounds like you know what it needs  to do and need the fix. Maybe best if you fix it.

**Comment by asvetlov** [comment]
Deal

**Comment by asvetlov** [comment]
Fixed by #2341
